### PR TITLE
[dynamo] add torch.compiler.set_stance

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137504

Attempt # 2 at https://github.com/pytorch/pytorch/pull/132926 to implement https://github.com/pytorch/pytorch/issues/123771.

Implement a new `torch.compiler.set_stance` function that can force `torch.compile` regions to run eagerly.

See added tests for usage examples.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/_dynamo/eval_frame.py
Comment (williamwen42) [inline torch/_dynamo/eval_frame.py:140]: Using manual context manager/decorator class instead of `contextlib.contextmanager` in order to mark a decorated function as forbidden in graph.

File: torch/_dynamo/eval_frame.py
Comment (williamwen42) [reply torch/_dynamo/eval_frame.py:137]: Flag is set, e.g. here https://github.com/pytorch/pytorch/blob/38027b9b4776f3fc3034193b28b7f80b58afb5de/torch/_dynamo/decorators.py#L168 to implement `forbid_in_graph`.

It is checked here https://github.com/pytorch/pytorch/blob/38027b9b4776f3fc3034193b28b7f80b58afb5de/torch/_dynamo/trace_rules.py#L3085

Comment (yf225) [reply torch/_dynamo/eval_frame.py:137]: ah I see, thanks!

**Review by yf225 — APPROVED** [review]

File: torch/compiler/__init__.py
Comment (jansel) [inline torch/compiler/__init__.py:232]: `phase` seems like the wrong name here.  Forcing eager does not seem like a compilation phase.

There also uses of this that don't have anything to do with phases.

File: torch/csrc/dynamo/eval_frame.c
Comment (jansel) [inline torch/csrc/dynamo/eval_frame.c:834]: Doesn't this do the same thing as `set_eval_frame(None)` (which disables dynamo and makes everything run eagerly).

We also have `TORCHDYNAMO_DISABLE=1` which does something similar.

The one advantage this has is it is thread local, but IMO we should just fix the thread safety of the existing stuff rather than adding a duplicate API.

**Review by jansel — CHANGES_REQUESTED** [review]

**Comment by jansel** [comment]
Also note that a version of `eager_on_recompile` already exists with `torch._dynamo.run` (maps to `set_eval_frame(True)`), so no need to recreate that.

Similarly, `fail_on_recompile` can be done by setting the backend to one that raises an error (without any changes to eval_frame.c).

File: torch/csrc/dynamo/eval_frame.c
Comment (williamwen42) [reply torch/csrc/dynamo/eval_frame.c:834]: `eval_frame_callback_enabled_get` is higher priority than `set_eval_frame`. The new API is useful for cases where we want to ignore future `torch.compile` invocations, which would overwrite the eval frame handler.

We have a lot of discussion over at https://github.com/pytorch/pytorch/pull/132926.

Comment (jansel) [reply torch/csrc/dynamo/eval_frame.c:834]: It would be much faster to disable frame evaluation entirely (do the check once when we hit torch.compiled code), since then the CPython JIT can kick in.  We also won't need to do an extra check on every function call.

**Comment by ezyang** [comment]
I'm going to defer to @jansel on this since he has more opinions on the internal implementation details, but holler if you need me to unblock you

File: torch/_dynamo/decorators.py
Comment (jansel) [inline torch/_dynamo/decorators.py:96]: Won't calling this in the constructor mean that, defining a function:
```
@set_stance("foo")
def bar():
   pass
```
Will change the global stance even if `bar()` is unused?

File: torch/_dynamo/eval_frame.py
Comment (jansel) [inline torch/_dynamo/eval_frame.py:147]: Add a comment about why this is needed.

File: torch/compiler/__init__.py
Comment (jansel) [inline torch/compiler/__init__.py:242]: This needs to be updated.  Let's also improve it with some example usage.

File: torch/_dynamo/eval_frame.py
Comment (jansel) [inline torch/_dynamo/eval_frame.py:123]: Should we assert here that there isn't currently an eval frame hook set?

If there is, then the stance won't take effect until the next dynamo region is entered.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/decorators.py
Comment (williamwen42) [reply torch/_dynamo/decorators.py:96]: `__call__` will run in decorator mode, which resets the stance state to the previous state (this logic was taken from the `torch.set_grad_enabled` implementation).

File: torch/_dynamo/eval_frame.py
Comment (williamwen42) [reply torch/_dynamo/eval_frame.py:123]: `_dynamo_forbidden` prevents `set_stance` from being used in a `torch.compile` region. Was there another case you were thinking about?

Comment (jansel) [reply torch/_dynamo/eval_frame.py:130]: Calling this function after a graph break that forces a fallback to eager.

File: torch/_dynamo/eval_frame.py
Comment (jansel) [inline torch/_dynamo/eval_frame.py:153]: We could do this check inside `set_stance` (since that is run less often than this one)

**Review by jansel — APPROVED** [review]

**Comment by williamwen42** [comment]
@ezyang it's already implemented in this PR, although I overlooked adding a test and including it as a usage example. I can add this in a followup.

**Comment by yf225** [comment]
@williamwen42 also I've started using the direct-mutating API: https://github.com/pytorch/pytorch/pull/138113/files and it's awesome :)

**Pull request closed without merging**
